server/bookmarks.go

```diff
@@ -0,0 +1,220 @@
+	if err != nil {
+		return err
+	}
+
+	bookmarkList, err := core.GetBookmarkList(r.ctx, s.db, listID)
+	if err != nil {
+		return err
+	}
+
+	if bookmarkList.UserID != *r.viewer {
+		return errForbidden
+	}
+
+	bookmarks, err := core.GetBookmarksFromList(r.ctx, s.db, listID)
+	if err != nil {
+		return err
+	}
+
+	if len(bookmarks) == 0 {
+		return w.writeJSON([]struct{}{})
+	}
+
+	return w.writeJSON(bookmarks)
+}
+
+// /api/bookmarks/:listID [DELETE]
+func (s *Server) deleteBookmarkList(w *responseWriter, r *request) error {
+	if !r.loggedIn {
+		return errNotLoggedIn
+	}
+
+	listID, err := strToID(r.muxVar("listID"))
+	if err != nil {
+		return err
+	}
+
+	bookmarkList, err := core.GetBookmarkList(r.ctx, s.db, listID)
+	if err != nil {
+		return err
+	}
+
+	if bookmarkList.UserID != *r.viewer {
+		return errForbidden
+	}
+
+	err = core.DeleteBookmarkList(r.ctx, s.db, listID)
+	if err != nil {
+		return err
+	}
+
+	return w.writeJSON(nil)
+
+}
+
+// /api/bookmarks/:listID [POST]
+func (s *Server) addBookmark(w *responseWriter, r *request) error {
+	if !r.loggedIn {
+		return errNotLoggedIn
+	}
+
+	listID, err := strToID(r.muxVar("listID"))
+	if err != nil {
+		return err
+	}
+
+	bookmarkList, err := core.GetBookmarkList(r.ctx, s.db, listID)
+	if err != nil {
+		return err
+	}
+
+	if bookmarkList.UserID != *r.viewer {
+		return errForbidden
+	}
+
+	req := struct {
+		ItemType string `json:"itemType"`
+		ItemID   uid.ID `json:"itemId"`
+	}{}
+
+	if err := r.unmarshalJSONBody(&req); err != nil {
+		return err
+	}
+
+	// Ensure that ItemType exists and is either "post" or "comment"
+	if req.ItemType != "post" && req.ItemType != "comment" {
+		return nil
+	}
+
+	bookmark, err := core.CreateBookmark(r.ctx, s.db, listID, req.ItemType, req.ItemID)
+	if err != nil {
+		return err
+	}
+
+	return w.writeJSON(bookmark)
+}
+
+// /api/bookmarks/:listID/:itemID [GET]
+func (s *Server) getBookmark(w *responseWriter, r *request) error {
+	if !r.loggedIn {
+		return errNotLoggedIn
+	}
+
+	listID, err := strToID(r.muxVar("listID"))
+	if err != nil {
+		return err
+	}
+
+	itemID, err := strToID(r.muxVar("itemID"))
+	if err != nil {
+		return err
+	}
+
+	bookmarkList, err := core.GetBookmarkList(r.ctx, s.db, listID)
+	if err != nil {
+		return err
+	}
+
+	if bookmarkList.UserID != *r.viewer {
+		return errForbidden
+	}
+
+	bookmark, err := core.GetBookmark(r.ctx, s.db, itemID, listID)
+	if err != nil {
+		return err
+	}
+
+	return w.writeJSON(bookmark)
+}
+
+// /api/bookmarks/:listID/:itemID [DELETE]
+func (s *Server) deleteBookmark(w *responseWriter, r *request) error {
+	if !r.loggedIn {
+		return errNotLoggedIn
+	}
+
+	listID, err := strToID(r.muxVar("listID"))
+	if err != nil {
+		return err
+	}
+
+	itemID, err := strToID(r.muxVar("itemID"))
+	if err != nil {
+		return err
+	}
+
+	bookmarkList, err := core.GetBookmarkList(r.ctx, s.db, listID)
+	if err != nil {
+		return err
+	}
+
+	if bookmarkList.UserID != *r.viewer {
+		return errForbidden
```

Comment:
What if instead of throwing forbidden we just do a 404? Both responses will work but 404 ensures that no one even knows that a bookmark list of that id even exists.

Reply:
I thought this originally, actually! Not sure why I ended up going with a 401